### PR TITLE
Introduce tiered HNSW index MOD-4300 MOD-4299

### PR DESCRIPTION
**Pull request by alonre24**
- Introduce abstract `VecSimTieredIndex` that inherits from `VecSimIndexInterface`, and its `TieredHNSWIndex` implementation. 
- Add an API in HNSW factory that creates `TieredHNSWIndex` based on combined params (for HNSW and for the async tiered index properties) 
- `VecSimTieredIndex` contains a pointer to BF index, and creates it as part of the object constructor, based on the main index params (this had required adding two getters to `VecSimIndexAbstract`).
- Add the required data structures and callbacks to the common defines (that is, the "job" definition, submit to queue callback etc...)
- Add a new unit test suit, that uses "mock" job queue and callbacks to allow testing of the proper creation of the  `TieredHNSWIndex` (to be used later on in tests...)

**Notes**
- This PR include upgrading our build to c++20 standard (to be suitable with `<shared_mutex>` object)
- All the pure virtual `VecSimIndexInterface` functions were to be implemented in order to create an instance. These functions (and some other signature functions) are (temporary) not covered by tests.

### CODE REVIEW

File: src/VecSim/vec_sim_tiered_index.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+class VecSimTieredIndex : public VecSimIndexInterface {
+protected:
+    BruteForceIndex<DataType, DistType> *tempFlat;
+    VecSimIndexAbstract<DistType> *index;
+
+    void *jobQueue;
+    SubmitCB SubmitJobsToQueue;
+
+    void *memoryCtx;
+    UpdateMemoryCB UpdateIndexMemory;
```
Comment (GuyAv46) [inline src/VecSim/vec_sim_tiered_index.h:18]: Consider documenting what are the expectations from these callbacks and pointers (here or with the type definitions)

File: src/VecSim/algorithms/hnsw/hnsw_factory.cpp [hunk 2]
```diff
@@ -119,4 +120,22 @@ size_t EstimateElementSize(const HNSWParams *params) {
...
+    if (params->hnswParams.type == VecSimType_FLOAT32) {
+        auto *hnsw_index =
+            NewIndex_ChooseMultiOrSingle<float, float>(&params->hnswParams, allocator);
+        return new (allocator) TieredHNSWIndex<float, float>(
+            hnsw_index, params->jobQueue, params->submitCb, params->memoryCtx, params->UpdateMemCb);
+    } else if (params->hnswParams.type == VecSimType_FLOAT64) {
+        auto *hnsw_index =
+            NewIndex_ChooseMultiOrSingle<double, double>(&params->hnswParams, allocator);
+        return new (allocator) TieredHNSWIndex<double, double>(
+            hnsw_index, params->jobQueue, params->submitCb, params->memoryCtx, params->UpdateMemCb);
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_factory.cpp:135]: Why not just pass `&params->hnswParams` to the hnsw new function? this way we'll have to update both this and the HNSW new function when support for other types will be added

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+                    void *memory_ctx, UpdateMemoryCB UpdateMemCb)
+        : VecSimTieredIndex<DataType, DistType>(hnsw_index, job_queue_, submitCb, memory_ctx,
+                                                UpdateMemCb) {}
+    virtual ~TieredHNSWIndex() = default;
+
+    // TODO: Implement the actual methods instead of these temporary ones.
+    int addVector(const void *blob, labelType id) override {
+        return this->index->addVector(blob, id);
+    }
+    int deleteVector(labelType id) override { return this->index->deleteVector(id); }
```
Comment (GuyAv46) [inline src/VecSim/algorithms/hnsw/hnsw_tiered.h:39]: So long for async...

**Review by GuyAv46 — COMMENTED** [review]
Looks very good!
Consider reorganizing the hnsw directory into subdirectories (tiered, batch...)

File: src/VecSim/vec_sim_common.h [hunk 4]
```diff
@@ -83,11 +90,30 @@ typedef struct {
...
+    void *memoryCtx;
+    UpdateMemoryCB UpdateMemCb;
+} TieredHNSWParams;
+
+typedef enum { HNSW_INSERT_JOB, HNSW_DELETE_JOB } HNSWJobType;
+
+typedef struct HNSWJob {
+    HNSWJobType jobType;
+    labelType label;
+    void *blob;
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/vec_sim_common.h:106]: I guess... we can adjust this struct later on for multi if we see that we'll need to...

File: src/VecSim/vec_sim_tiered_index.h (hunk 1 above)
Comment (alonre24) [reply src/VecSim/vec_sim_tiered_index.h:18]: I added docus in `vec_sim_common.h`

**Review by GuyAv46 — APPROVED** [review]
LGTM

File: src/VecSim/vec_sim_common.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/vec_sim_common.h:110]: After revising, the insert job struct should contain the internal id as well, so we can access each blob in a MULTI index for moving it from the flat into the HNSW index as part of every separate job.

File: src/VecSim/vec_sim_common.h [hunk 5]
```diff
@@ -83,11 +90,34 @@ typedef struct {
...
+    UpdateMemoryCB
+        UpdateMemCb; // A callback that updates the memoryCtx with a given memory (number).
+} TieredHNSWParams;
+
+typedef enum { HNSW_INSERT_JOB, HNSW_DELETE_JOB } HNSWJobType;
+
+/**
+ * Definition of HNSW job structure for asynchronous tiered index.
+ */
+typedef struct HNSWJob {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/vec_sim_common.h:107]: Changed to generic

File: src/VecSim/vec_sim_tiered_index.h [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+public:
+    VecSimTieredIndex(VecSimIndexAbstract<DistType> *index_, void *job_queue_, SubmitCB submitCb,
+                      void *memory_ctx, UpdateMemoryCB UpdateMemCb)
+        : VecSimIndexInterface(index_->getAllocator()), index(index_), jobQueue(job_queue_),
+          SubmitJobsToQueue(submitCb), memoryCtx(memory_ctx), UpdateIndexMemory(UpdateMemCb) {
+        BFParams bf_params = {.type = index_->getType(),
+                              .dim = index_->getDim(),
+                              .metric = index_->getMetric(),
+                              .multi = index_->isMultiValue()};
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/vec_sim_tiered_index.h:33]: So you propose a change, or just wanted to raise awareness?

File: src/VecSim/algorithms/hnsw/hnsw_tiered.h [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+typedef struct UpdateMessage {
+    HNSWJobType type;
+    // ...
+} UpdateMessage;
+
+template <typename DataType, typename DistType>
+class TieredHNSWIndex : public VecSimTieredIndex<DataType, DistType> {
+private:
+    // Todo: implement these methods later on
+    void executeJob(HNSWJob *job) {}
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_tiered.h:20]: It has changed in the meantime, but the intention is to implement every specific job's callback directly in the derived async index that uses that job...

File: src/VecSim/algorithms/brute_force/brute_force.h [hunk 8]
```diff
@@ -16,6 +16,7 @@
 #include <queue>
 #include <cassert>
 #include <limits>
+#include <stdexcept>
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/algorithms/brute_force/brute_force.h:19]: redundant...

File: src/VecSim/vec_sim_common.h [hunk 9]
```diff
@@ -92,14 +99,80 @@ typedef struct {
     size_t blockSize;
 } BFParams;
 
+// A struct that contains the common tiered index params.
+typedef struct {
+    void *jobQueue;    // External queue that holds the jobs.
+    SubmitCB submitCb; // A callback that submits an array of jobs into a given jobQueue.
+    void *memoryCtx;   // External context that stores the index memory consumption.
+    UpdateMemoryCB
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_common.h:107]: is this the result of `make format`?

File: src/VecSim/vec_sim_tiered_index.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+#include "vec_sim_index.h"
+#include "algorithms/brute_force/brute_force.h"
+
+#include <shared_mutex>
+
+template <typename DataType, typename DistType>
+class VecSimTieredIndex : public VecSimIndexInterface {
+protected:
+    BruteForceIndex<DataType, DistType> *tempFlat;
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_tiered_index.h:11]: maybe consider rename to `flatBuffer`

File: src/VecSim/algorithms/hnsw/hnsw_factory.h [hunk 11]
```diff
@@ -16,6 +16,8 @@
 namespace HNSWFactory {
 
 VecSimIndex *NewIndex(const HNSWParams *params, std::shared_ptr<VecSimAllocator> allocator);
+VecSimIndex *NewTieredIndex(const TieredHNSWParams *params,
```
Comment (DvirDukhan) [inline src/VecSim/algorithms/hnsw/hnsw_factory.h:19]: who calls this function right now other than the tests?

File: src/VecSim/vec_sim_tiered_index.h [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+    void *memoryCtx;
+    UpdateMemoryCB UpdateIndexMemory;
+
+    // Consider putting these in the derived class instead. Also - see if we should use
+    // std::shared_mutex
+    std::shared_timed_mutex flatIndexGuard;
+    std::shared_timed_mutex mainIndexGuard;
+
+public:
+    VecSimTieredIndex(VecSimIndexAbstract<DistType> *index_, void *job_queue_, SubmitCB submitCb,
```
[the remark this answers is not in the document]
Comment (DvirDukhan) [reply src/VecSim/vec_sim_tiered_index.h:26]: this comment still relevant

File: tests/unit/test_hnsw_tiered.cpp [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+        VecSimIndex_AddVector(tiered_index->tempFlat, vector, vector_label);
+
+        // Create a mock job that inserts some vector into the HNSW index.
+        auto insert_to_index = [](void *job) {
+            auto *my_insert_job = reinterpret_cast<HNSWInsertJob *>(job);
+            auto my_index =
+                reinterpret_cast<TieredHNSWIndex<TEST_DATA_T, TEST_DIST_T> *>(my_insert_job->index);
+
+            // Move the vector from the temp flat index into the HNSW index.
+            my_index->addVector(my_index->tempFlat->getDataByInternalId(my_insert_job->id),
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_tiered.cpp:44]: add comment explaining we are addign by internal id due to multi value index

File: tests/unit/test_hnsw_tiered.cpp [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+                                my_insert_job->label);
+            VecSimIndex_DeleteVector(my_index->tempFlat, my_insert_job->label);
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_tiered.cpp:46]: delete by internal id - new PR

File: tests/unit/test_hnsw_tiered.cpp [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+            auto it = my_index->labelToInsertJobs[my_insert_job->label].begin();
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_tiered.cpp:47]: explain this flow happens in a fixture that expects this task already be in the queue
in other words
this fixture tests the `labelToInsertJobs` correctness

File: tests/unit/test_hnsw_tiered.cpp [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+            my_index->labelToInsertJobs[my_insert_job->label].erase(it);
+            my_index->UpdateIndexMemory(my_index->memoryCtx,
+                                        my_index->getAllocator()->getAllocationSize());
+        };
+
+        HNSWInsertJob job = {
+            .base = AsyncJob{.jobType = HNSW_INSERT_VECTOR_JOB, .Execute = insert_to_index},
+            .index = tiered_index,
+            .label = vector_label};
+        tiered_index->labelToInsertJobs[vector_label].push_back(&job);
```
Comment (DvirDukhan) [inline tests/unit/test_hnsw_tiered.cpp:58]: don't we need a specific API call for this, not breaking abstraction? 
maybe add TODO to push this logic into the `add_vector` of the tiered ineex

File: src/VecSim/vec_sim_common.h (hunk 9 above)
Comment (alonre24) [reply src/VecSim/vec_sim_common.h:107]: I guess... I'll reorganize it

File: src/VecSim/vec_sim_common.h [hunk 17]
```diff
@@ -92,14 +99,80 @@ typedef struct {
...
         HNSWParams hnswParams;
         BFParams bfParams;
+        TieredHNSWParams tieredHNSWParams;
     };
 } VecSimParams;
 
+/**
+ * The specific job types in use (to be extended in the future by demand)
+ */
+typedef enum {
```
[the remark this answers is not in the document]
Comment (alonre24) [reply src/VecSim/vec_sim_common.h:128]: The idea was to put all the definitions in this PR, and split the implementation into multiple tasks...
I guess that we can use `UNDEFINED` or `INVALID_JOB` for the readability of future functions that will check what is the job type.

File: src/VecSim/algorithms/hnsw/hnsw_factory.h (hunk 11 above)
Comment (alonre24) [reply src/VecSim/algorithms/hnsw/hnsw_factory.h:19]: Currently no one. In the future, it should be called from `VecSimIndex_New`. Should it be part of this PR?

File: src/VecSim/vec_sim_common.h [hunk 18]
```diff
@@ -129,7 +129,8 @@ typedef enum {
     HNSW_INSERT_VECTOR_JOB,
     HNSW_REPAIR_NODE_CONNECTIONS_JOB,
     HNSW_SEARCH_JOB,
-    HNSW_SWAP_JOB
+    HNSW_SWAP_JOB,
+    INVALID_JOB // to indicate that finding a JobType >= INVALID_JOB is an error
```
Comment (DvirDukhan) [inline src/VecSim/vec_sim_common.h:133]: I think this should be the first one

File: tests/unit/CMakeLists.txt [hunk 19]
```diff
@@ -13,8 +13,8 @@ include(CTest)
 
 set(CMAKE_CXX_STANDARD 20)
 
-set(CMAKE_CXX_FLAGS_DEBUG "-g")
-set(CMAKE_CXX_FLAGS_RELEASE "-O3")
+set(CMAKE_CXX_FLAGS_DEBUG "-g -pthread")
+set(CMAKE_CXX_FLAGS_RELEASE "-O3 -pthread")
```
Comment (DvirDukhan) [inline tests/unit/CMakeLists.txt:17]: are we using pthreads here?

File: src/VecSim/vec_sim_common.h (hunk 18 above)
Comment (alonre24) [reply src/VecSim/vec_sim_common.h:133]: why?

File: tests/unit/CMakeLists.txt (hunk 19 above)
Comment (alonre24) [reply tests/unit/CMakeLists.txt:17]: I thought it was necessary for using the `Thread` object in the test but looks like it doesn't. removing...

**Review by DvirDukhan — APPROVED** [review]

**Pull request merged**